cupy/core/core.pyx

```diff
@@ -2890,6 +2893,151 @@ cpdef ndarray tensordot_core(
+    # compute C^T = B^T * A here.
+    a, transa, lda = _mat_to_cublas_contiguous(a, 0)
+    b, transb, ldb = _mat_to_cublas_contiguous(b, 1)
+
+    if out_dtype in 'efF':
+        compute_type = cublas.CUBLAS_COMPUTE_32F
+    elif out_dtype in 'dD':
+        compute_type = cublas.CUBLAS_COMPUTE_64F
+
+    compute_capability = int(device.get_compute_capability())
+    algo = cublas.CUBLAS_GEMM_DEFAULT
+    if ((compute_capability >= 80) or
+            (compute_capability >= 70 and out_dtype == 'e')):
+        algo = cublas.CUBLAS_GEMM_DEFAULT_TENSOR_OP
+
+    handle = device.get_cublas_handle()
+    a_cuda_dtype = _get_cuda_dtype(a)
+    b_cuda_dtype = _get_cuda_dtype(b)
+    c_cuda_dtype = _get_cuda_dtype(c)
+    if out_dtype in 'efd':
+        if compute_type == cublas.CUBLAS_COMPUTE_32F:
+            one_f = 1
+            zero_f = 0
+            cublas.gemmEx(
+                handle, <int>transb, <int>transa, <int>m, <int>n, <int>k,
+                <size_t>&one_f, b.data.ptr, b_cuda_dtype, <int>ldb,
+                a.data.ptr, a_cuda_dtype, <int>lda,
+                <size_t>&zero_f, c.data.ptr, c_cuda_dtype, <int>m,
+                compute_type, algo)
+        elif compute_type == cublas.CUBLAS_COMPUTE_64F:
+            one_d = 1
+            zero_d = 0
+            cublas.gemmEx(
+                handle, <int>transb, <int>transa, <int>m, <int>n, <int>k,
+                <size_t>&one_d, b.data.ptr, b_cuda_dtype, <int>ldb,
+                a.data.ptr, a_cuda_dtype, <int>lda,
+                <size_t>&zero_d, c.data.ptr, c_cuda_dtype, <int>m,
+                compute_type, algo)
+        else:
+            raise ValueError('Invalid compute type: {}'.format(compute_type))
+    elif out_dtype in 'FD':
+        if compute_type == cublas.CUBLAS_COMPUTE_32F:
+            one_F = cuComplex(1, 0)
+            zero_F = cuComplex(0, 0)
+            cublas.gemmEx(
+                handle, <int>transb, <int>transa, <int>m, <int>n, <int>k,
+                <size_t>&one_F, b.data.ptr, b_cuda_dtype, <int>ldb,
+                a.data.ptr, a_cuda_dtype, <int>lda,
+                <size_t>&zero_F, c.data.ptr, c_cuda_dtype, <int>m,
+                compute_type, algo)
+        elif compute_type == cublas.CUBLAS_COMPUTE_64F:
+            one_D = cuDoubleComplex(1, 0)
+            zero_D = cuDoubleComplex(0, 0)
+            cublas.gemmEx(
+                handle, <int>transb, <int>transa, <int>m, <int>n, <int>k,
+                <size_t>&one_D, b.data.ptr, b_cuda_dtype, <int>ldb,
+                a.data.ptr, a_cuda_dtype, <int>lda,
+                <size_t>&zero_D, c.data.ptr, c_cuda_dtype, <int>m,
+                compute_type, algo)
+        else:
+            raise ValueError('Invalid compute type: {}'.format(compute_type))
+    else:
+        raise ValueError('Invalid dtype: %s' % str(out_dtype))
+
+    if out is not ret:
+        elementwise_copy(out, ret)
+    return ret
+
+
+cdef int _get_cuda_dtype(ndarray a):
```

Comment:
1. Need to propagate exception if it's raised
2. Compare char directly
```suggestion
cdef int _get_cuda_dtype(ndarray a) except -1:
    cdef str a_type = a.dtype.char
```

Reply:
I'm wondering if this function should go to `cupy/core/_dtype.pyx` instead...?

Reply:
Yeah I think a little refactoring would be great: note that the very same function is also needed in cuSPARSE and cuTENSOR, for example:
https://github.com/cupy/cupy/blob/ca79633dbeb3bbf95678957535249e398db74aca/cupy/cusparse.py#L70-L80
How about modifying the signature like this:
```python
cdef int _dtype_to_cuda_type(dtype, is_half_allowed=False) except -1
```
and reuse it everywhere in the codebase?

Reply:
Agreed. I also think it's better to reuse a function that convert from numpy data types to CUDA data types. I'd like to see the following implementation, what would you think on this?
```
cpdef int dtype_to_cuda_dtype(dtype_char, available_dtype_char=None) except -1:
    if available_dtype_char is None:
        available_dtype_char = 'fdFD'
    if dtype_char not in available_dtype_char:
        raise TypeError('dtype is not available: %s' % str(dtype_char))
    if dtype_char == 'e':
        return runtime.CUDA_R_16F
    elif dtype_char == 'f':
        return runtime.CUDA_R_32F
    elif dtype_char == 'd':
        return runtime.CUDA_R_64F
    elif dtype_char == 'F':
        return runtime.CUDA_C_32F
    elif dtype_char == 'D':
        return runtime.CUDA_C_64F
    else:
        raise TypeError('dtype is not supported: %s' % str(dtype_char))
```

Reply:
Hi @anaruse My preference is to keep the NumPy dtype as input, because when raising an error it offers a better description than a single char. Also, we can avoid double comparison (your first `not in` and then the `if`'s). Last, `available_dtype_char` is useless because the if branches are limited. 

I think this could be simpler: 
```python
cpdef int dtype_to_cuda_dtype(dtype, bint is_half_allowed=False) except -1:
    cdef str dtype_char = dtype.char

    if dtype_char == 'e' and is_half_allowed:
        return runtime.CUDA_R_16F
    elif dtype_char == 'f':
        return runtime.CUDA_R_32F
    elif dtype_char == 'd':
        return runtime.CUDA_R_64F
    elif dtype_char == 'F':
        return runtime.CUDA_C_32F
    elif dtype_char == 'D':
        return runtime.CUDA_C_64F
    else:
        raise TypeError('dtype is not supported: {}'.format(dtype))
```